library/ssl_srv.c

```diff
@@ -2499,6 +2612,61 @@ static void ssl_write_alpn_ext( mbedtls_ssl_context *ssl,
 }
 #endif /* MBEDTLS_ECDH_C || MBEDTLS_ECDSA_C */
 
+#if defined(MBEDTLS_SSL_DTLS_SRTP ) && defined(MBEDTLS_SSL_PROTO_DTLS)
+static void ssl_write_use_srtp_ext( mbedtls_ssl_context *ssl,
+                                    unsigned char *buf,
+                                    size_t *olen )
+{
+    size_t mki_len = 0, ext_len = 0;
+    uint16_t profile_value = 0;
+
+    if( ssl->dtls_srtp_info.chosen_dtls_srtp_profile == MBEDTLS_SRTP_UNSET_PROFILE )
+    {
+        *olen = 0;
+        return;
+    }
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "server hello, adding use_srtp extension" ) );
+
+    if( ssl->conf->dtls_srtp_mki_support == MBEDTLS_SSL_DTLS_SRTP_MKI_SUPPORTED &&
+        ssl->dtls_srtp_info.mki_len != 0 )
+    {
+        mki_len = ssl->dtls_srtp_info.mki_len;
+    }
+
+    /* extension */
+    buf[0] = (unsigned char)( ( MBEDTLS_TLS_EXT_USE_SRTP >> 8 ) & 0xFF );
+    buf[1] = (unsigned char)( ( MBEDTLS_TLS_EXT_USE_SRTP      ) & 0xFF );
+    /*
+     * total length 5 and mki value: only one profile(2 bytes)
+     *              and length(2 bytes) and srtp_mki  )
+     */
+    ext_len = 5 + mki_len;
+    buf[2] = (unsigned char)( ( ext_len >> 8 ) & 0xFF );
+    buf[3] = (unsigned char)( ext_len & 0xFF );
+
+    /* protection profile length: 2 */
+    buf[4] = 0x00;
+    buf[5] = 0x02;
+    profile_value = mbedtls_ssl_get_srtp_profile_iana_value( ssl->dtls_srtp_info.chosen_dtls_srtp_profile );
+    if( profile_value != 0xFFFF )
+    {
+        buf[6] = (unsigned char)( ( profile_value >> 8 ) & 0xFF );
+        buf[7] = (unsigned char)( profile_value & 0xFF );
+    }
+    else
+    {
+        *olen = 0;
+        return;
+    }
+
+    buf[8] = mki_len & 0xFF;
+    memcpy( &buf[9], ssl->dtls_srtp_info.mki_value, mki_len );
```

Comment:
Blocker: It's a deficiency of our current extension-writing interface that we don't pass an end-of-buffer field to bounds-check the extension writing functions against. That's normally not an issue for statically sized extensions, assuming that Mbed TLS isn't configured with extremely small buffer sizes, but in the case at hand, we're echoing the client-provided MKI value which has variable length and thereby has the potential to silently overflow the write buffer. 

Suggestion: Add an `end` pointer to the extension writing function and add an explicit bounds check.

Reply:
This one is a bit harder than the others. All the extension writing functions shall be rewritten.
I can't see in the ssl_write_server_hello function a way to access the writing buffer size. Is there a constant that define this buffer size somewhere?
The client hello extension is variable but not unbounded, its max size is 255 bytes.

Reply:
@hanno-arm any hint to get the actual max buffer size?

Reply:
Unless I'm mistaken, this days one should use `ssl->out_msg + MBEDTLS_SSL_OUT_CONTENT_LEN` to get the end of the buffer (or more precisely, the limit usable for record contents - the actual buffer is a bit larger in order to allow for record protection overhead).

Reply:
Thanks @mpg I added the check on buffer size as done on the `ssl_write_cid_ext` using  `ssl->out_msg + MBEDTLS_SSL_OUT_CONTENT_LEN` as you suggested.